File: internal/domain/entity/time_slot.go
package entity

import (
	"beauty-server/internal/domain/errors"
	"beauty-server/internal/domain/value_object"
	"github.com/google/uuid"
	"time"
)

type TimeSlot struct {
	Id        uuid.UUID
	StaffId   uuid.UUID
	VenueId   uuid.UUID
	Weekday   time.Weekday
	Intervals []value_object.Interval
}

type VenueRepository interface {
	GetById(id uuid.UUID) (*Venue, error)
}

func NewTimeSlot(id, staffId, venueId uuid.UUID, weekday time.Weekday, venueRepository VenueRepository) (*TimeSlot, error) {
	venue, err := venueRepository.GetById(venueId)
	if err != nil {
		return nil, err
	}
	if venue == nil {
		return nil, errors.NewErrVenueNotFound(venueId)
	}

	return &TimeSlot{
		Id:        id,
		StaffId:   staffId,
		VenueId:   venue.Id,
		Weekday:   weekday,
		Intervals: make([]value_object.Interval, 0),
	}, nil
}

func (ts TimeSlot) UpdateIntervals(intervals []value_object.Interval) error {
	for i := 0; i < len(intervals); i++ {
		for j := i + 1; j < len(intervals); j++ {
			if intervals[i].Overlaps(intervals[j]) {
				return errors.NewErrIntervalsOverlap(
					intervals[i].Start, intervals[i].End,
					intervals[j].Start, intervals[j].End,
				)
			}
		}
	}

	ts.Intervals = intervals
	return nil
}

File: Domain/Entities/TimeSlot.cs
using Domain.Errors;
using Domain.Primitives;
using Domain.Repositories.Venues;
using Domain.Shared;
using Domain.ValueObjects;

namespace Domain.Entities;

public class TimeSlot : Entity
{
    private TimeSlot(
        Guid id,
        Guid staffId,
        Guid venueId,
        DateOnly date,
        List<Interval> intervals) : base(id)
    {
        StaffId = staffId;
        VenueId = venueId;
        Date = date;
        Intervals = intervals;
    }
    
#pragma warning disable CS8618
    private TimeSlot() { }
#pragma warning restore CS8618

    public Guid StaffId { get; private set; }
    public Guid VenueId { get; private set; }
    public DateOnly Date { get; private set; }
    public List<Interval> Intervals { get; private set; }

    public static async Task<Result<TimeSlot>> CreateAsync(
        Guid id,
        Guid staffId,
        Guid venueId,
        DateOnly date,
        IVenueReadOnlyRepository venueRepository,
        CancellationToken cancellationToken = default)
    {
        var venueExists = await venueRepository.ExistsAsync(venueId, cancellationToken);
        if (!venueExists)
            return Result.Failure<TimeSlot>(DomainErrors.Venue.NotFound(venueId));

        return new TimeSlot(id, staffId, venueId, date, new List<Interval>());
    }
    
    public Result AddInterval(TimeSpan start, TimeSpan end)
    {
        var intervalResult = Interval.Create(start, end);
        if (intervalResult.IsFailure)
            return intervalResult;

        var newInterval = intervalResult.Value;

        foreach (var existingInterval in Intervals)
        {
            if (existingInterval.Overlaps(newInterval))
            {
                return Result.Failure(DomainErrors.TimeSlot.IntervalsOverlap);
            }
        }

        Intervals.Add(newInterval);
        return Result.Success();
    }


    public Result UpdateIntervals(List<Interval> intervals)
    {
        for (int i = 0; i < intervals.Count; i++)
        {
            for (int j = i + 1; j < intervals.Count; j++)
            {
                if (intervals[i].Overlaps(intervals[j]))
                {
                    return Result.Failure(DomainErrors.TimeSlot.IntervalsOverlap);
                }
            }
        }

        Intervals = intervals;
        return Result.Success();
    }
}